src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningConditionMapper.java

```diff
@@ -0,0 +1,367 @@
+import com.salesforce.dynamodbv2.mt.mappers.sharedtable.impl.HashPartitioningKeyMapper.PhysicalRangeKeyBytesConverter;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.antlr.v4.runtime.BailErrorStrategy;
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+
+class HashPartitioningConditionMapper implements ConditionMapper {
+
+    private static final byte[] EMPTY_BYTE_ARRAY = new byte[] {};
+
+    private final HashPartitioningKeyMapper keyMapper;
+
+    HashPartitioningConditionMapper(HashPartitioningKeyMapper keyMapper) {
+        this.keyMapper = keyMapper;
+    }
+
+    @Override
+    public void applyForUpdate(UpdateItemRequest updateItemRequest) {
+        // TODO
+    }
+
+    @Override
+    public void applyToFilterExpression(RequestWrapper request) {
+        // no-op, since we keep all original virtual fields
+    }
+
+    @Override
+    public void applyToKeyCondition(RequestWrapper request, RequestIndex requestIndex) {
+        // parse expression for the virtual HK value, and the RK condition if it exists
+        ExpressionsParser parser = getExpressionsParser(request.getExpression());
+        KeyConditionExpressionVisitor visitor = new KeyConditionExpressionVisitor(requestIndex.getVirtualPk(),
+            request.getExpressionAttributeNames(), request.getExpressionAttributeValues());
+        parser.keyConditionExpression().accept(visitor);
+        checkNotNull(visitor.hkValue, "Key condition does not specify hash key");
+
+        // map virtual expression & values to physical expression & values
+        KeyConditionExpressionMapper expressionMapper = new KeyConditionExpressionMapper(requestIndex.getVirtualPk(),
+            requestIndex.getPhysicalPk(), visitor.hkValue, visitor.rkCondition);
+        String physicalExpression = expressionMapper.mapKeyConditionExpression(request);
+        request.setExpression(physicalExpression);
+    }
+
+    /**
+     * Maps a virtual query key condition expression to a physical one. Summary:
+     * <ul>
+     *     <li>"#vhk = :vhk", no vrk ==> "#prk = :vhk“
+     *     <li>"#vhk = :vhk and #vrk = :vrk" ==> "#prk = :vhk+vrk“
+     *     <li>"#vhk = :vhk", has vrk ==> "#prk BETWEEN :vhk AND :vhk+FFFFFF)"
+     *     <li>"#vhk = :vhk and #vrk < :vrk" ==> "#prk BETWEEN :vhk AND :vhk+(vrk-1)+FFFFFF“
+     *     <li>"#vhk = :vhk and #vrk <= :vrk" ==> "#prk BETWEEN :vhk AND :vhk+vrk“
+     *     <li>"#vhk = :vhk and #vrk > :vrk" ==> "#prk BETWEEN :vhk+vrk+0 AND :vhk+FFFFFF“
+     *     <li>"#vhk = :vhk and #vrk >= :vrk" ==> "#prk BETWEEN :vhk+vrk AND :vhk+FFFFFF“
+     *     <li>"#vhk = :vhk and #vrk BETWEEN :vrk1 AND :vrk2" ==> "#prk BETWEEN :vhk+vrk1 AND :vhk+vrk2“
+     * </ul>
+     */
```

Comment:
The main query key condition expression mapping logic is here

Reply:
Great javadoc, thanks!